### ISSUE TITLE
Request to Expose AudioBuffer to DedicatedWorker

### ISSUE THREAD

**Issue by chcunningham** [WebAudio/web-audio-api#2288]
WebCodecs uses AudioBuffer as a container of unencoded audio.
https://wicg.github.io/web-codecs/#dom-audioframe-buffer

WebCodecs AudioDecoder and AudioEncoder classes operate on AudioBuffer and are exposed to DedicatedWorker to facilitate offloaded codec IO.

I think the change is as simple as changing the Exposed= parameter for just AudioBuffer. I'll send a PR to demonstrate.

**Comment by padenot** [comment]
We need to reason about the (somewhat) hidden cost of the interleaving/deinterleaving operations that we'll face, and the cost of the possible (probable) memory inflation that this might cause.

Generally, decoding, encoding and IO APIs are using interleaved audio buffer, so that temporal locality is better: samples that are close in time are close in memory (they are played out, recorded, encoded and decoded together). Processing APIs often like deinterleaved (also sometimes called planar by analogy with the graphics world) buffers, because processing a particular audio channel with a particular algorithm is more efficient/practical this way.

The Web Audio API always use planar audio, because its main goal is processing. It happens to have a decoding API for historical reasons, that is made to inter-operate with the rest of the API, and this contradicts the above.

Additionally, the Web Audio API always uses 32-bits floating point numbers to represent audio samples, when exposed to authors. We often see codecs output (and input) in 16-bits integers (or less, or more, but 16-bits is the most common). This is a 2-fold increased in memory footprint for no real good reasons, that led Firefox for Android to have a lazy inflation scheme, where audio samples stay in 16-bits integer (what the decoder provides) until they need to be exposed to script, this works well but is implicit. https://github.com/WebAudio/web-audio-api-v2/issues/11 aims at making this explicit and allow further optimizations.

It might well be that we're okay with this, and then implementation do those inflation/reordering operations in a lazy fashion. We could also do a spin-off of https://github.com/WebAudio/web-audio-api-v2/issues/11 and allow interleaved audio, alongside different sample-representation. This seems like a better way to do things, and doesn't seem particularly hard (no harder than what is implemented today in Gecko).

**Comment by hoch** [comment]
I think this is about exposing AudioBuffer to DedicatedWorker, not to AudioWorkletGlobalScope. Paul, is your concern mostly about the inflated memory footprint?

**Comment by padenot** [comment]
And CPU, memory traffic, etc., yes.

Basically, unnecessary inefficiencies at the API boundary because the type that exists today is not what a codec API needs.

Just exposing it is fine, it's not hard. Indeed, this has nothing to do with `AudioWorkletGlobalScope`.

**Comment by hoch** [comment]
I agree with that we need to look into the hidden implication, but exposing AudioBuffer itself seems harmless and it would be useful for WebCodec's use cases.
